main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -0,0 +1,852 @@
+      val name = method.getName
+      val descriptor = asm.Type.getMethodDescriptor(method)
+      // If the method is static, use INVOKESTATIC otherwise use INVOKEVIRTUAL
+      val invokeInsn = if (Modifier.isStatic(method.getModifiers)) INVOKESTATIC else INVOKEVIRTUAL
+      visitor.visitMethodInsn(invokeInsn, declaration, name, descriptor, false)
+      // If the method is void, put a unit on top of the stack
+      if (asm.Type.getType(method.getReturnType) == asm.Type.VOID_TYPE) {
+        visitor.visitMethodInsn(INVOKESTATIC, JvmName.Unit.toInternalName, "getInstance",
+          AsmOps.getMethodDescriptor(List(), JvmType.Unit), false)
+      }
+
+    case Expression.UserError(_, loc) =>
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      val msg = s"User exception: ${loc.format}."
+      compileException(visitor, JvmName.UserException, msg)
+
+    case Expression.MatchError(_, loc) =>
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      val msg = s"Non-exhaustive match expression: ${loc.format}."
+      compileException(visitor, JvmName.MatchException, msg)
+
+    case Expression.SwitchError(_, loc) =>
+      // Adding source line number for debugging
+      addSourceLine(visitor, loc)
+      val msg = s"Non-exhaustive switch expression: ${loc.format}."
+      compileException(visitor, JvmName.SwitchException, msg)
+  }
+
+  private def compileException(visitor: MethodVisitor, className: JvmName, msg: String): Unit = {
+    visitor.visitTypeInsn(NEW, className.toInternalName)
+    visitor.visitInsn(DUP)
+    visitor.visitLdcInsn(msg)
+    // TODO: Load actual source location or change the exception
+    visitor.visitFieldInsn(GETSTATIC, "ca/uwaterloo/flix/language/ast/package$SourceLocation$", "MODULE$", "Lca/uwaterloo/flix/language/ast/package$SourceLocation$;")
+    visitor.visitMethodInsn(INVOKEVIRTUAL, "ca/uwaterloo/flix/language/ast/package$SourceLocation$", "Unknown", "()Lca/uwaterloo/flix/language/ast/package$SourceLocation;", false)
+    visitor.visitMethodInsn(INVOKESPECIAL, className.toInternalName, "<init>", "(Ljava/lang/String;Lca/uwaterloo/flix/language/ast/package$SourceLocation;)V", false)
+    visitor.visitInsn(ATHROW)
+  }
+
+  /*
+   * Generate code to load an integer constant.
+   *
+   * Uses the smallest number of bytes necessary, e.g. ICONST_0 takes 1 byte to load a 0, but BIPUSH 7 takes 2 bytes to
+   * load a 7, and SIPUSH 200 takes 3 bytes to load a 200. However, note that values on the stack normally take up 4
+   * bytes. The exception is if we set `isLong` to true, in which case a cast will be performed if necessary.
+   *
+   * This is needed because sometimes we expect the operands to be a long, which means two (int) values are popped from
+   * the stack and concatenated to form a long.
+   */
+  private def compileInt(visitor: MethodVisitor)(i: Long, isLong: Boolean = false): Unit = {
```

Comment:
I know the originally code had this use of currying, but there is no need for it.